tensordict/nn/common.py

```diff
@@ -771,29 +777,81 @@ def reset_out_keys(self):
                 del self._forward_hooks[i]
         return self
 
-    def reset_parameters(self):
+    def reset_parameters_recursive(
+        self, parameters: Optional[TensorDictBase] = None
+    ) -> Optional[TensorDictBase]:
         """Recursively reset the parameters of the module and its children.
 
+        Args:
+            parameters (TensorDict of parameters, optional): If set to None, the module will reset using self.parameters().
+                Otherwise, we will reset the parameters in the tensordict in-place. This is
+                useful for functional modules where the parameters are not stored in the module itself.
+
+        Returns:
+            A tensordict of the new parameters, only if parameters was not None.
+
         Examples:
-        >>> from tensordict.nn import TensorDictModule
-        >>> from torch import nn
-        >>> net = nn.Sequential(nn.Linear(2,3), nn.ReLU())
-        >>> old_param = net[0].weight.clone()
-        >>> module = TensorDictModule(net, in_keys=['bork'], out_keys=['dork'])
-        >>> module.reset_parameters()
-        >>> (old_param == net[0].weight).all()
-        tensor(False)
+            >>> from tensordict.nn import TensorDictModule
+            >>> from torch import nn
+            >>> net = nn.Sequential(nn.Linear(2,3), nn.ReLU())
+            >>> old_param = net[0].weight.clone()
+            >>> module = TensorDictModule(net, in_keys=['bork'], out_keys=['dork'])
+            >>> module.reset_parameters()
+            >>> (old_param == net[0].weight).any()
+            tensor(False)
+
+            This method also supports functional parameter sampling:
+
+            >>> from tensordict import TensorDict
+            >>> from tensordict.nn import TensorDictModule
+            >>> from torch import nn
+            >>> net = nn.Sequential(nn.Linear(2,3), nn.ReLU())
+            >>> module = TensorDictModule(net, in_keys=['bork'], out_keys=['dork'])
+            >>> params = TensorDict.from_module(module)
+            >>> old_params = params.clone(recurse=True)
+            >>> module.reset_parameters(params)
+            >>> (old_params == params).any()
+            False
         """
-        for m in self.children():
-            self._reset_parameters(m)
+        if parameters is None:
+            self._reset_parameters(self)
+            return
```

Comment:
Can we also test this? (for usage with vmap)

```suggestion
            return
        elif parameters.ndim:
            for param in parameters.unbind(0):
                self.reset_parameters_recursive(self, param)
            return
```